File: wrappers/math/test/src/TestMath.cxx
/*
 * Copyright (C) 2007 Robotics at Maryland
 * Copyright (C) 2007 Joseph Lisee <jlisee@umd.edu>
 * All rights reserved.
 *
 * Author: Joseph Lisee <jlisee@umd.edu>
 * File:  wrappers/math/test/src/TestMath.cxx
 */

// STD Includes
#include <cmath>

// Library Includes
#include <UnitTest++/UnitTest++.h>
#include <boost/bind.hpp>
#include <boost/function.hpp>
#include <boost/shared_ptr.hpp>
#include <boost/python.hpp>

namespace py = boost::python;

struct MathFixture
{
    MathFixture() :
        main_module(py::import("__main__")),
        main_namespace(main_module.attr("__dict__")),
        eval(boost::bind(py::exec, _1, main_namespace, main_namespace))
    {
        main_namespace["math"] = py::import("ext.math");
    }

    py::object main_module;
    py::object main_namespace;
    boost::function<py::object (py::str)> eval;
};

TEST(MathImport)
{
    py::import("ext.math");
}

TEST_FIXTURE(MathFixture, Radian)
{
    try {
        eval("d = math.Degree(45)\n"
             "deg = d.valueDegrees()\n"
             "rad    = d.valueRadians()\n");
        
        double value = py::extract<double>(main_namespace["deg"]);
        CHECK_EQUAL(45, value);
        value = py::extract<double>(main_namespace["rad"]);
        CHECK_CLOSE(M_PI/4, value, 0.1);
        
        
        eval("import math as m\n"
             "r = math.Radian(m.pi)\n"
             "deg = r.valueDegrees()\n"
             "rad = r.valueRadians()\n");
    
        value = py::extract<double>(main_namespace["deg"]);
        CHECK_CLOSE(180, value, 0.1);
        value = py::extract<double>(main_namespace["rad"]);
        CHECK_CLOSE(M_PI, value, 0.1);


    } catch(py::error_already_set err) { PyErr_Print(); throw err; }
}

TEST_FIXTURE(MathFixture, Matrix3)
{
    try  {
        eval("M = math.Matrix3(1,2,3,4,5,6,7,8,9)\n"
             "elem21 = M[2][1]");
        double value = py::extract<double>(main_namespace["elem21"]);
        CHECK_EQUAL(8, value);
        
    } catch(py::error_already_set err) { PyErr_Print(); throw err; }
}

TEST_FIXTURE(MathFixture, __str__)
{
    try {
        eval("vset = str(math.Vector3(1,2,3))");

        std::string str = py::extract<std::string>(main_namespace["vset"]);
        CHECK_EQUAL("Vector3(1, 2, 3)", str);
    } catch(py::error_already_set err) { PyErr_Print(); throw err; }
}

int main()
{
    // We must register this contverted so we can pass a pointer to python
//    py::register_ptr_to_python< boost::shared_ptr<ram::control::IController> >();
    Py_Initialize();
    return UnitTest::RunAllTests();
}

File: wrappers/math/test/src/TestMath.py
# Copyright (C) 2008 Maryland Robotics Club
# Copyright (C) 2008 Joseph Lisee <jlisee@umd.edu>
# All rights reserved.
#
# Author: Joseph Lisee <jlisee@umd.edu>
# File: wrapper/math/test/src/TestMath.py

# STD Imports
import unittest
import math as pmath

# Project Imports
import ext.math as math

class TestConversions(unittest.TestCase):
    def testRadianDegree(self):
        self.assertEquals(20, math.Degree(20).valueDegrees())

        rad = math.Radian(math.Degree(20))
        self.assertEquals(20, rad.valueDegrees())

        deg = math.Degree(math.Radian(pmath.pi))
        self.assertAlmostEqual(180, deg.valueDegrees(), 3)

        
        # Ensure no automatic double to Radian/Degree conversions occur
        try:
            deg - 1
            self.fail("Implicit conversions from float/int not allowed")
        except TypeError:
            pass

        try:
            deg - 2.0
            self.fail("Implicit conversions from float/int not allowed")
        except TypeError:
            pass
        
        try:
            rad - 3
            self.fail("Implicit conversions from float/int not allowed")
        except TypeError:
            pass

        try:
            rad - 4.0
            self.fail("Implicit conversions from float/int not allowed")
        except TypeError:
            pass

class TestVectorMatrix2(unittest.TestCase):
    def test(self):
        start = math.Vector2(1,0);
        rotationMatrix = math.Matrix2()
        rotationMatrix.fromAngle(math.Degree(45));
        
        expected = math.Vector2(math.Vector2(1,1).normalisedCopy());
        # TODO: Expose the free operator functions
        #result = (start * rotationMatrix)

        #self.assertAlmostEqual(expected.x, result.x, 4)
        #self.assertAlmostEqual(expected.y, result.y, 4)

if __name__ == '__main__':
    unittest.main()
